federation-2/router-bridge/js-src/plan.ts

```diff
@@ -216,6 +216,53 @@ export class BridgeQueryPlanner {
       return { data, errors: [] };
     }
   }
+
+  operationSignature(
+    operationString: string,
+    providedOperationName?: string
+  ): string {
+    let document: DocumentNode;
+
+    try {
+      document = parse(operationString);
+    } catch (parseError) {
+      // parse throws GraphQLError
+      return PARSE_FAILURE;
+    }
+
+    // Federation does some validation, but not all.  We need to do
+    // all default validations that are provided by GraphQL.
+    const validationErrors = validate(this.apiSchema, document);
+    if (validationErrors.length > 0) {
+      return VALIDATION_FAILURE;
+    }
+
+    let operation: Operation;
+    try {
+      operation = operationFromDocument(this.composedSchema, document, {
+        operationName: providedOperationName,
+      });
+    } catch (e) {
+      return VALIDATION_FAILURE;
+    }
+
+    // Adapted from here
+    // https://github.com/apollographql/apollo-server/blob/444c403011209023b5d3b5162b8fb81991046b23/packages/apollo-server-core/src/requestPipeline.ts#L303
+    const operationName = operation?.name;
+
+    // I double checked, this function doesn't throw
+    const operationDerivedData = getOperationDerivedData({
+      schema: this.apiSchema,
+      document,
+      operationName,
+    });
+
+    const statsReportKey = `# ${operationName || "-"}\n${
+      operationDerivedData.signature
+    }`;
+
+    return statsReportKey;
+  }
```

Comment:
I factored in https://github.com/apollographql/federation-rs/pull/350/commits/e4ec69a1fa21e4452137524e47a3c54be96d1780

operation and usage reporting generation are better done in one phase, since the usage reporting structure carries the error